src/audio/mixin_mixout/mixin_mixout_generic.c

```diff
@@ -0,0 +1,417 @@
+		n = MIN(samples, nmax);
+		nmax = audio_stream_samples_without_wrap_s24(sink, dst);
+		n = MIN(n, nmax);
+		/* frames is the processed frame count in this loop*/
+		frames = 0;
+		for (i = 0; i < n; i += source_channel_count) {
+			*dst = sat_int24(sign_extend_s24(*dst) +
+					  (int32_t)q_mults_32x32(sign_extend_s24(*src),
+					  gain, IPC4_MIXIN_GAIN_SHIFT));
+			src += source_channel_count;
+			dst += sink_channel_count;
+			frames++;
+		}
+	}
+
+	for (left_frames = frames_to_copy; left_frames > 0; left_frames -= frames) {
+		src = audio_stream_wrap(source, src);
+		dst = audio_stream_wrap(sink, dst);
+		nmax = audio_stream_samples_without_wrap_s24(source, src);
+		samples = left_frames * source_channel_count;
+		n = MIN(samples, nmax);
+		nmax = audio_stream_samples_without_wrap_s24(sink, dst);
+		n = MIN(n, nmax);
+		frames = 0;
+		for (i = 0; i < n; i += source_channel_count) {
+			*dst = (int32_t)q_mults_32x32(sign_extend_s24(*src),
+							   gain, IPC4_MIXIN_GAIN_SHIFT);
+			src += source_channel_count;
+			dst += sink_channel_count;
+			frames++;
+		}
+	}
+}
+
+#endif	/* CONFIG_FORMAT_S24LE */
+
+#if CONFIG_FORMAT_S32LE
+/* Instead of using sink->channels and source->channels, sink_channel_count and
+ * source_channel_count are supplied as parameters. This is done to reuse the function
+ * to also mix an entire stream. In this case the function is called with fake stream
+ * parameters: multichannel stream is treated as single channel and so the entire stream
+ * contents is mixed.
+ */
+static void normal_mix_channel_s32(struct audio_stream __sparse_cache *sink, uint32_t start_frame,
+				   uint32_t mixed_frames,
+				   const struct audio_stream __sparse_cache *source,
+				   uint32_t frame_count, uint16_t gain)
+{
+	uint32_t frames_to_mix, frames_to_copy, left_frames;
+	uint32_t n, nmax, i;
+	int32_t *dst = (int32_t *)sink->w_ptr + start_frame;
+	int32_t *src = (int32_t *)source->r_ptr;
+
+	assert(mixed_frames >= start_frame);
+	frames_to_mix = mixed_frames - start_frame;
+	frames_to_mix = MIN(frames_to_mix, frame_count);
+	frames_to_copy = frame_count - frames_to_mix;
+
+	for (left_frames = frames_to_mix; left_frames > 0; left_frames -= n) {
+		src = audio_stream_wrap(source, src);
+		dst = audio_stream_wrap(sink, dst);
+		/* calculate the remaining samples*/
+		nmax = audio_stream_samples_without_wrap_s32(source, src);
+		n = MIN(left_frames, nmax);
+		nmax = audio_stream_samples_without_wrap_s32(sink, dst);
+		n = MIN(n, nmax);
+		for (i = 0; i < n; i++) {
+			*dst = sat_int32((int64_t)*dst + (int64_t)*src++);
+			dst++;
+		}
+	}
+
+	for (left_frames = frames_to_copy; left_frames > 0; left_frames -= n) {
+		src = audio_stream_wrap(source, src);
+		dst = audio_stream_wrap(sink, dst);
+		nmax = audio_stream_samples_without_wrap_s32(source, src);
+		n = MIN(left_frames, nmax);
+		nmax = audio_stream_samples_without_wrap_s32(sink, dst);
+		n = MIN(n, nmax);
+		for (i = 0; i < n; i++)
+			*dst++ = *src++;
```

Comment:
do you think a memory copy will be faster here?

Reply:
Good catch, it's very strange to see `*dst++ = *src++;` in a PR named "Optimize"

https://docs.zephyrproject.org/apidoc/latest/string_8h.html
https://github.com/thesofproject/sof/blob/main/src/arch/xtensa/include/arch/string.h
https://github.com/thesofproject/sof/blob/main/zephyr/include/rtos/string.h

Reply:
test on xtensa simulator and my TGL several times, use __vec_memcpy to replace the L/S performace worse!
So just change generic C to use memcpy_s， the latest results are:
for 16bit normal mode mix, optimized C version can save about 56.9% cycles;
for 24bit normal mode mix, optimized C version can save about 62.9% cycles;
for 32bit normal mode mix, optimized C version can save about 55.5% cycles.

Reply:
> use __vec_memcpy to replace the L/S performace worse!

This looks like a very serious bug. Please file a bug and add a link to it here in the source.